contracts/service-registry/src/state.rs

```diff
@@ -130,16 +131,73 @@ pub enum AuthorizationState {
     Authorized,
 }
 
-// maps service_name -> Service
-pub const SERVICES: Map<&str, Service> = Map::new("services");
-// maps (service_name, chain_name, worker_address) -> ()
-pub const WORKERS_PER_CHAIN: Map<(&str, &ChainName, &Addr), ()> = Map::new("workers_per_chain");
-// maps (service_name, worker_address) -> Worker
-pub const WORKERS: Map<(&str, &Addr), Worker> = Map::new("workers");
+type ChainNames = HashSet<ChainName>;
+type ServiceName = str;
+type WorkerAddress = Addr;
+
+pub const SERVICES: Map<&ServiceName, Service> = Map::new("services");
+pub const WORKERS_PER_CHAIN: Map<(&ServiceName, &ChainName, &WorkerAddress), ()> =
+    Map::new("workers_per_chain");
+pub const CHAINS_PER_WORKER: Map<(&ServiceName, &WorkerAddress), ChainNames> =
+    Map::new("chains_per_worker");
+pub const WORKERS: Map<(&ServiceName, &WorkerAddress), Worker> = Map::new("workers");
+
+pub fn register_chain_support_utility(
+    storage: &mut dyn Storage,
+    service_name: String,
+    chains: Vec<ChainName>,
+    caller: WorkerAddress,
+) -> Result<Response, ContractError> {
+    let mut worker_current_chains = CHAINS_PER_WORKER
+        .may_load(storage, (&service_name, &caller))?
+        .unwrap_or_else(HashSet::new);
+
+    for chain in chains {
+        WORKERS_PER_CHAIN.save(storage, (&service_name, &chain, &caller), &())?;
+        worker_current_chains.insert(chain.clone());
+    }
+
+    CHAINS_PER_WORKER.save(storage, (&service_name, &caller), &(worker_current_chains))?;
+
+    Ok(Response::new())
+}
```

Comment:
```suggestion
pub fn register_chain_support(
    storage: &mut dyn Storage,
    service_name: String,
    chains: Vec<ChainName>,
    caller: WorkerAddress,
) -> Result<(), ContractError> {
    CHAINS_PER_WORKER.update(storage, (&service_name, &caller), |current_chains| {
        let mut current_chains = current_chains.unwrap_or_else(HashSet::new);
        current_chains.extend(chains.iter());
        Ok(current_chains)
    })?;

    for chain in chains.iter() {
        WORKERS_PER_CHAIN.save(storage, (&service_name, chain, &caller), &())?;
    }

    Ok(())
}
```

Reply:
the state doesn't need to know about `Response`

Reply:
thank you very much for the code!